src/api/v1/reports.ts

```diff
@@ -0,0 +1,107 @@
+import { Hono } from "hono";
+import * as vocab from "@fedify/fedify/vocab";
+
+import { z } from "zod";
+import { db } from "../../db";
+import { zValidator } from "@hono/zod-validator";
+
+import federation from "../../federation";
+import { type Variables, scopeRequired, tokenRequired } from "../../oauth";
+import { accounts, posts, reports, type Post, type Report } from "../../schema";
+import { uuidv7 } from "uuidv7-js";
+import { serializeReport } from "../../entities/report";
+import { eq, inArray } from "drizzle-orm";
+
+const app = new Hono<{ Variables: Variables }>();
+
+const reportSchema = z.object({
+  comment: z.string().trim().min(1).max(1000).optional(),
+  account_id: z.string().uuid(),
+  status_ids: z.array(z.string().uuid()).min(1).optional(),
+  // discarded by defined by the Mastodon API:
+  category: z.string().optional(),
+  rule_ids: z.string().array().optional(),
+  forward: z.boolean().optional()
+});
+
+app.post(
+  "/",
+  tokenRequired,
+  scopeRequired(["write:reports"]),
+  zValidator(
+    "json",
+    reportSchema
+  ),
+  async (c) => {
+    const accountOwner = c.get("token").accountOwner;
+    if (accountOwner == null) {
+      return c.json(
+        { error: "This method requires an authenticated user" },
+        422,
+      );
+    }
+
+    const data = c.req.valid("json");
+
+    const targetAccount = await db.query.accounts.findFirst({
+      where: eq(accounts.id, data.account_id),
+      with: { owner: true, successor: true }
+    });
```

Comment:
There doesn't seem to be anything in the database to indicate whether or not an Account is local to this server or a remote server?

Reply:
If `targetAccount.owner` is `null` it's remote. Otherwise it's local.

Reply:
So like:

```
where: and(eq(account.id, data.account_id), isNull(account.owner))
```

Should work? I'm not sure that would given the database layout — the accounts table doesn't seem to have an `owner` column, so I think that'd need some sort of join instead?

Reply:
`account_owners.id` references `account.id`.  So it should be like:

```js
where: and(
  eq(account.id, data.account_od),
  notInArray(account.id, db.select({ id: accountOwners.id }).from(accountOwners))
)
```

Reply:
Okay, have done that.